snappy/desktop.go

```diff
@@ -0,0 +1,164 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snappy
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/ubuntu-core/snappy/dirs"
+	"github.com/ubuntu-core/snappy/osutil"
+)
+
+// this is a bit confusing, a localestring in xdg is:
+// "key=value" or "key[locale]=" with locale as:
+//  lang_COUNTRY@MODFIER (or a subset of this)
+var desktopFileI18nPattern = `(|\[[a-zA-Z_@]+\])`
+var validDesktopFileLines = []*regexp.Regexp{
+	// headers
+	regexp.MustCompile(`^\[Desktop Entry\]$`),
+	regexp.MustCompile(`^\[Desktop Action [a-zA-Z0-9]+\]$`),
+	// whitespace lines
+	regexp.MustCompile(`^\s*$`),
+	// lines with comments
+	regexp.MustCompile(`^\s*#`),
+	// https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s05.html
+	regexp.MustCompile(`^Type=`),
+	regexp.MustCompile(`^Version=`),
+	regexp.MustCompile(fmt.Sprintf(`^Name%s=`, desktopFileI18nPattern)),
+	regexp.MustCompile(fmt.Sprintf(`^GenericName%s=`, desktopFileI18nPattern)),
+	regexp.MustCompile(`^NoDisplay=`),
+	regexp.MustCompile(fmt.Sprintf(`^Comment%s=`, desktopFileI18nPattern)),
+	regexp.MustCompile(`^Icon=`),
+	regexp.MustCompile(`^Hidden=`),
+	regexp.MustCompile(`^OnlyShowIn=`),
+	regexp.MustCompile(`^NotShowIn=`),
+	regexp.MustCompile(`^Exec=`),
+	// Note that we do not support TryExec, it does not make sense
+	// in the snap context
+	regexp.MustCompile(`^Terminal=`),
+	regexp.MustCompile(`^Actions=`),
+	regexp.MustCompile(`^MimeType=`),
+	regexp.MustCompile(`^Categories=`),
+	regexp.MustCompile(fmt.Sprintf(`^Keywords%s=`, desktopFileI18nPattern)),
+	regexp.MustCompile(`^StartupNotify=`),
+	regexp.MustCompile(`^StartupWMClass`),
+}
+
+func isValidDesktopFilePrefix(line string) bool {
+	for _, re := range validDesktopFileLines {
+		if re.MatchString(line) {
+			return true
+		}
+	}
+	return false
+}
+
+func rewriteExecLine(m *snapYaml, line string) (string, error) {
+	cmd := strings.SplitN(line, "=", 2)[1]
+	for _, app := range m.Apps {
+		validCmd := filepath.Base(generateBinaryName(m, app))
+		// check the prefix to allow %flag style args
+		// this is ok because desktop files are not run through sh
+		// so we don't have to worry about the arguments too much
```

Comment:
I'm not sure that's a valid assumption. I don't know that we can reasonably protect against it, but we should probably at least change the comment to "Unity 7 doesn't execute them with a shell, and that's all we care about." There are a lot of desktops out there that do different things.